theano/gof/utils.py

```diff
@@ -216,13 +217,37 @@ def __call__(cls, *args, **kwargs):
         props = cls.__dict__.get('__props__', None)
         if props is None:
             return super(MetaObject, cls).__call__(*args, **kwargs)
-        key = (cls, args, tuple(kwargs.items()))
-        try:
-            if key not in cls._cache.keys():
-                cls._cache[key] = super(MetaObject, cls).__call__(*args, **kwargs)
-            return cls._cache[key]
-        except:
-            return super(MetaObject, cls).__call__(*args, **kwargs)
+        key = [cls]
+        for arg in args:
+            arg = make_hashable(arg)
+            key.append(arg)
+        for k, v in kwargs.iteritems():
+            key.append(k)
+            v = make_hashable(v)
+            key.append(v)
+        key = tuple(key)
+        if key not in cls._cache:
+            cls._cache[key] = super(MetaObject, cls).__call__(*args, **kwargs)
+        return cls._cache[key]
+
+
+def make_hashable(argument):
+    try:
+        hash(argument)
+        return argument
+    except:
+        if isinstance(argument, list) or isinstance(argument, tuple):
+            argument = list(argument)
+            for index, instance in enumerate(argument):
+                argument[index] = make_hashable(instance)
+            argument = tuple(argument)
+        elif isinstance(argument, dict):
+            argument = frozendict(argument)
+            for k, v in argument.iteritems():
+                v = make_hashable(v)
```

Comment:
Yes, I thought it would generate lots of errors! So I decided to first try this and debug this part and then change the values after the tests were passed.